src/rust/base/src/lib.rs

```diff
@@ -0,0 +1,68 @@
+// Copyright 2019 Shift Cryptosecurity AG
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+#![no_std]
+
+use arrayvec::ArrayString;
+use core::fmt::Write;
+use core::panic::PanicInfo;
+use core::time::Duration;
+
+mod screen;
+mod workflow;
+mod config;
+
+mod error {
+    #[derive(Debug)]
+    pub enum Error {
+        InvalidHostname,
+    }
+}
+
+#[panic_handler]
+fn panic(info: &PanicInfo) -> ! {
+    let mut buf = ArrayString::<[_; 256]>::new();
+    let _ = write!(buf, "Internal error: {}", info);
+    screen::print_debug(Duration::from_millis(0), &buf);
+    loop {}
+}
+
+#[no_mangle]
+pub extern fn workflow_display_base32(bytes: *const u8, bytes_len: usize) -> bool {
+    assert!(!bytes.is_null());
+    assert!(bytes_len > 0 && bytes_len <= 32);
+    let bytes = unsafe { core::slice::from_raw_parts(bytes, bytes_len) };
+    workflow::display_base32(bytes)
+}
+
+static mut CONFIG: config::Config = config::Config::new();
```

Comment:
is there no type inference for something like this?

also what is `CONFIG:` vs `config::Config`?

Reply:
`CONFIG` is the variable name `config::Config` is the type. Globals don't have type inference yet IIRC.